control-plane/src/e3_client.rs

```diff
@@ -0,0 +1,70 @@
+use std::net::SocketAddr;
+
+use tokio::net::TcpStream;
+
+use crate::{
+    error::Result,
+    internal_dns::{self, AsyncDnsResolver},
+};
+
+pub struct E3Client {
+    socket_addr: Option<SocketAddr>,
+    #[allow(unused)]
+    dns_resolver: AsyncDnsResolver,
+}
+
+impl Default for E3Client {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl E3Client {
+    pub fn new() -> Self {
+        let dns_resolver = internal_dns::get_internal_dns_resolver()
+            .expect("Failed to create internal dns resolver");
+
+        Self {
+            socket_addr: None,
+            dns_resolver,
+        }
+    }
+
+    pub async fn new_conn(&self) -> Result<TcpStream> {
+        let socket_addr = match self.socket_addr {
+            Some(addr) => addr,
+            None => self.get_ip_for_e3().await.unwrap(),
+        };
+
+        let conn = Self::try_connect(socket_addr).await;
+
+        if conn.is_ok() {
+            return Ok(conn.unwrap());
+        }
+
+        // try again with refreshed dns lookup
+        let socket_addr = self.get_ip_for_e3().await.expect("failed to get ip for e3");
```

Comment:
Do you need to store the new ip here?

Reply:
It would be worth looking at https://doc.rust-lang.org/std/sync/struct.RwLock.html as it allows many readers at once but only one writer